### ISSUE TITLE
Provider::estimate_gas does not set `from` field before estimating

### ISSUE THREAD

**Issue by sblOWPCKCR** [gakonst/ethers-rs#748]
**Version**
`ethers-providers v0.5.5`

**Platform**
`Darwin wild.local 21.2.0 Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000 arm64`

**Description**
`estimate_gas` fails to estimate gas for transactions that fail is msg.sender is 0x0. 

`estimate_gas` is just: `self.request("eth_estimateGas", [tx]).await`, so it expects the user to prepare the transaction **completely**. I don't think most of the people do that, because it's not required for `send_transaction`. 

`send_transaction` is:
```
        let mut tx = tx.into();
        self.fill_transaction(&mut tx, block).await?;
        self.request("eth_sendTransaction", [tx]).await?;
```
(and fill_transaction takes care of setting `from`).
Easy fix is to add identical `fill_transaction` call to `estimate_gas`

Be careful with reproducing this: it may not be easy to reproduce with hardhat. Here's the tx I'm estimating:
```
{
"jsonrpc": "2.0",
"id": 1,
"method": "eth_estimateGas",
"params": [
  {
   "to": "0x9A76B03fDdd21283158325f51213D36A1bdfC52B",
    "data": "0x77516039f5cf4b4c717f1d3f9c523d350000000000000000000000000000000000000000"
   }
  ]
}
```

It fails to estimate on infura/alchemy (expectedly!):
```
❯❯❯ curl --data @/tmp/p https://eth-mainnet.alchemyapi.io/v2/$KEY
{"jsonrpc": "2.0", "id": 1, "error": {"code": 3, "message": "execution reverted: ST", "data": "0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000025354000000000000000000000000000000000000000000000000000000000000"}}%
```
But a hardhat node (fresh forked from the same URL) is happy to estimate:
```
❯❯❯ curl --data @/tmp/p http://127.0.0.1:9545/
{"jsonrpc":"2.0","id":1,"result":"0x6ebb0"}
```

My guess is hardhat auto-populates `from` to one of the built-in accounts

**Comment by gakonst** [comment]
I wonder what the right way to call `fill_transaction` is. 

We could do either of the below:
1. `fn estimate_gas(tx: &mut TypedTransaction)`
2.`fn estimate_gas<T: Into<TypedTransaction> + Send + Sync>(tx: T)` and do `let mut tx = tx.into()` inside.

Then we'd call `fill_transaction(tx)` without any problem.

**Comment by sblOWPCKCR** [comment]
I like $2, but it's going to be brutal when ENS names or access lists are involved (they'd be recomputed twice if I do estimate_gas() + send())
Looks like $1 is the path of the least resistance.

A middle ground:
```rust
async fn estimate_gas<T: Into<TypedTransaction> + Send + Sync>(tx: T) -> Result<(U256, TypedTransaction), ProviderError> {
        let mut tx = tx.into();
        self.fill_transaction(&mut tx, block).await?;
        Ok(self.request("eth_estimateGas", [tx]).await, tx)
}
async fn send_filled_transaction(&self, tx: &TypedTransaction) {
   let tx_hash = self.request("eth_sendTransaction", [tx]).await?;
   Ok(PendingTransaction::new(tx_hash, self).interval(self.get_interval()))
}
```

So: normal users are not confused by estimate_gas() estimating TXs as if they were send from 0x0; power users don't lose performance by doing:
```rust
let (gas, filled_tx) = provider.estimate_gas(tx).await?;
provider.send_filled_transaction(filled_tx).await?
```

This feels like a bigger change though, even the codegen would need to be updated

**Comment by gakonst** [comment]
Yeah, agree with the assessment. Let's go with #1 in that case, and hope it doesn't break too many things. Worst case we can use interior mutability to avoid it.
